internal/infrastructure/ceph/rbd.go

```diff
@@ -267,6 +357,304 @@ func applyDiffDataRecords(
+			}
+			return fmt.Errorf("failed to read tag byte: %w", err)
+		}
+
+		switch tag {
+		case 'w', 'z': // UPDATED DATA or ZERO DATA
+			offset, err := readLE64(diffFileReader)
+			if err != nil {
+				if errors.Is(err, csumreader.ErrChecksumMismatch) {
+					return csumreader.ErrChecksumMismatch
+				}
+				return fmt.Errorf("failed to read offset: %w", err)
+			}
+			length, err := readLE64(diffFileReader)
+			if err != nil {
+				if errors.Is(err, csumreader.ErrChecksumMismatch) {
+					return csumreader.ErrChecksumMismatch
+				}
+				return fmt.Errorf("failed to read length: %w", err)
+			}
+
+			// Detect overlapping ranges
+			if prevOffset+prevLength > offset {
+				return fmt.Errorf("overlapping or unsorted ranges detected: previous (%d, %d), current (%d, %d)",
+					prevOffset, prevLength, offset, length)
+			}
+			prevOffset = offset
+			prevLength = length
+
+			// Skip zero-length records
+			if length == 0 {
+				continue
+			}
+
+			// Expand raw image file if necessary
+			if offset+length > uint64(fileSize) {
+				fileSize := int64(math.Ceil(float64(offset+length)/float64(expansionUnitSize)) * float64(expansionUnitSize))
+				if err := unix.Fallocate(int(rawFile.Fd()), 0, 0, fileSize); err != nil {
+					return fmt.Errorf("failed to fallocate raw image file: %w", err)
+				}
+			}
+
+			// Apply the diff chunk-by-chunk and determine which checksum chunks
+			// cover the range [offset, offset+length].
+			startChunkIndex := offset / rawChecksumChunkSize
+			endChunkIndex := (offset + length - 1) / rawChecksumChunkSize
+
+			for chunkIndex := startChunkIndex; chunkIndex <= endChunkIndex; chunkIndex++ {
+
+				// Compute the portion of this chunk that should be written.
+				chunkStartPos := chunkIndex * rawChecksumChunkSize
+				chunkEndPos := chunkStartPos + rawChecksumChunkSize
+				writeStartPos := max(offset, chunkStartPos)
+				writeEndPos := min(offset+length, chunkEndPos)
+				writeSize := writeEndPos - writeStartPos
+
+				if writeSize == rawChecksumChunkSize {
+					if err := processEntireChunkWrite(
+						rawFile,
+						rawChecksumFile,
+						diffFileReader,
+						tag,
+						chunkIndex,
+						chunkStartPos,
+						rawChecksumChunkSize,
+						csumZero,
+					); err != nil {
+						return err
+					}
+				} else {
+					if err := processPartialChunkWrite(
+						rawFile,
+						rawChecksumFile,
+						diffFileReader,
+						tag,
+						chunkIndex,
+						chunkStartPos,
+						writeStartPos,
+						writeSize,
+						rawChecksumChunkSize,
+					); err != nil {
+						return err
+					}
+				}
+			}
+
+		case 'e': // END
+			return nil
+
+		default:
+			return fmt.Errorf("unexpected tag: %c", tag)
+		}
+	}
+}
+
+func processEntireChunkWrite(
+	dstFile *os.File,
+	dstChecksumFile *os.File,
+	diffFileReader *bufio.Reader,
+	tag byte,
+	chunkIndex uint64,
+	chunkStartPos uint64,
+	rawChecksumChunkSize uint64,
+	csumZero uint64,
+) error {
+	var csumUpdated uint64
+	chunkData := make([]byte, rawChecksumChunkSize)
```

Comment:
We don't have use these variables for ZERO DATA, so let's define them in the UPDATED DATA if branch.